raiden-ts/src/path/epics.ts

```diff
@@ -1,48 +1,154 @@
+  pathFind,
+  pathFindFailed,
+  pathFound,
+  pfsListUpdated,
+  udcBalanceFetch,
+  udcBalanceFetchFailed,
+  udcBalanceUpdate,
+} from './actions';
 import { channelCanRoute, pfsInfo, pfsListInfo } from './utils';
-import { PathResults, Paths } from './types';
+import { IOU, LastIOUResults, PathResults, Paths, PFS } from './types';
+import { concat } from 'ethers/utils/bytes';
+import { Signer } from 'ethers';
+
+const makeIOU = (sender: Address, receiver: Address, chainId: UInt<32>, oneToNAddress: Address) =>
+  ({
+    sender: sender,
+    receiver: receiver,
+    chain_id: chainId,
+    amount: Zero as UInt<32>,
+    one_to_n_address: oneToNAddress,
+    expiration_block: Zero as UInt<32>,
+    signature: HashZero as Signature,
+  } as IOU);
+
+const updateIOU = (iou: IOU, price: UInt<32>) =>
+  ({
+    ...iou,
+    amount: iou.amount.add(price),
+  } as IOU);
+
+const signIOU$ = (iou: IOU, signer: Signer) =>
+  from(signer.signMessage(
+    concat([
+      encode(iou.one_to_n_address, 20),
+      encode(iou.chain_id, 32),
+      encode(MessageTypeId.IOU, 32),
+      encode(iou.sender, 20),
+      encode(iou.receiver, 20),
+      encode(iou.amount, 32),
+      encode(iou.expiration_block, 32),
+    ]),
+  ) as Promise<Signature>).pipe(map(signature => ({ ...iou, signature } as IOU)));
+
+const makeAndSignLastIOURequest$ = (sender: Address, receiver: Address, signer: Signer) =>
+  defer(() => {
+    const payload = {
+      sender,
+      receiver,
+      timestamp: new Date(Date.now()).toISOString().split('.')[0],
+    };
+
+    const timestamp = hexlify(toUtf8Bytes(payload.timestamp));
+    const messageHash = concat([
+      encode(payload.sender, 20),
+      encode(payload.receiver, 20),
+      encode(timestamp, hexDataLength(timestamp)),
+    ]);
+    return from(signer.signMessage(messageHash) as Promise<Signature>).pipe(
+      map(signature => ({
+        ...payload,
+        signature,
+      })),
+    );
+  });
+
+const prepareNextIOU$ = (
+  state: RaidenState,
+  pfs: PFS,
+  deps: RaidenEpicDeps,
+  httpTimeout: number,
+  tokenNetwork: Address,
+) => {
+  const cachedIOU: IOU = get(state.iou, [tokenNetwork, pfs.address]);
+  return (cachedIOU
+    ? of(cachedIOU)
+    : makeAndSignLastIOURequest$(state.address, pfs.address, deps.signer).pipe(
+        mergeMap(payload =>
+          fromFetch(
+            `${pfs.url}/api/v1/${tokenNetwork}/payment/iou?${new URLSearchParams(
+              payload,
+            ).toString()}`,
+            {
+              method: 'GET',
+              headers: { 'Content-Type': 'application/json' },
+            },
+          ),
+        ),
+        timeout(httpTimeout),
+        mergeMap(async response => {
+          const text = await response.text();
+          if (response.status === 404) {
+            return makeIOU(
+              state.address,
+              pfs.address,
+              bigNumberify(state.chainId) as UInt<32>,
+              deps.contractsInfo.OneToN.address,
```

Comment:
Here's where a memoized async getter over `UserDeposit` contract is better than an explicit `contractsInfo` entry